generator/src/main/java/com/scottlogic/deg/generator/decisiontree/ConstraintNode.java

```diff
@@ -99,13 +99,26 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         ConstraintNode that = (ConstraintNode) o;
-        return Objects.equals(atomicConstraints, that.atomicConstraints) &&
-            Objects.equals(decisions, that.decisions);
+
+        boolean atomicConstraintsEqual = atomicConstraints.containsAll(that.atomicConstraints) &&
+                that.atomicConstraints.containsAll(atomicConstraints);
+        boolean delayedAtomicConstraintsEqual = delayedAtomicConstraints.containsAll(that.delayedAtomicConstraints) &&
+            that.delayedAtomicConstraints.containsAll(delayedAtomicConstraints);
+        boolean decisionsEqual = decisions.containsAll(that.decisions) &&
+            that.decisions.containsAll(decisions);
+
+        return atomicConstraintsEqual &&
+            delayedAtomicConstraintsEqual &&
+            decisionsEqual;
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(atomicConstraints, decisions);
+        List<AtomicConstraint> atomicConstraintsList = new ArrayList<>(atomicConstraints);
```

Comment:
Is there a more sane way of doing this?

Reply:
why are you creating array list collection wrappers round the fields that hashing?  looks like builder already using array list.

Reply:
This whole class is a bit smelly.  Eg why nodeMarkings not being used in equals or hashCode.  maybe nodeMarkinsg does not belong on this class if its not used in the equals.  Also adaptedRowSpec - It should not be stored in this class - lazy creating it and caching it in this class can only cause problems and makes it hard to use intellij generated equals and hashcode fns.

Reply:
I have to create the ArrayLists as a hash on Collections does not hash the objects within the Collection but a List does. And I'm unable to cast the Collection to a List.

I agree there does seem to be issues with this class generally. I suppect people have added properties to the class and not updated the Equals or Hash functions.

Reply:
I will raise a separate issue concerned with the flaws of this class (that existed before this PR too) - i'll link it to the underlying issue and/or this PR